pal init: Init membarrier() before first thread to improve start time

harisokanovic opened the pull request:
pal init: InitializeFlushProcessWriteBuffers() before first thread to improve start time

InitializeFlushProcessWriteBuffers() initializes expedited membarrier()
syscall on Linux, which is much slower when called in a multi-thread
process. Move this init earlier to improve dotnet process start time.
A detailed explanation can be found in issue 106722.

Fixes https://github.com/dotnet/runtime/issues/106722

jkotas commented on src/coreclr/pal/src/thread/process.cpp:
```diff
@@ -2575,20 +2605,11 @@ BOOL
-    int mask = membarrier(MEMBARRIER_CMD_QUERY, 0);
-    if (mask >= 0 &&
-        mask & MEMBARRIER_CMD_PRIVATE_EXPEDITED)
+    InitializeMembarrier();
```
Instead of splitting the initialization into two methods, could you please move the call to the existing `InitializeFlushProcessWriteBuffers` method to somewhere around here: https://github.com/dotnet/runtime/pull/106724/files#diff-ed6f693a713e751b280574475591c567e6adfbdb97b02a814411c0de8b9ae2e0R388

```
if (flags & PAL_INITIALIZE_FLUSH_PROCES_WRITE_BUFFERS)
{
    if (!InitializeFlushProcessWriteBuffers())
    {
        palError = ERROR_PALINIT_INITIALIZE_FLUSH_PROCESS_WRITE_BUFFERS;
        goto CLEANUP0a;
    }
}
```

+ define `PAL_INITIALIZE_FLUSH_PROCES_WRITE_BUFFERS` in pal.h:

`#define PAL_INITIALIZE_REGISTER_ACTIVATION_SIGNAL   0x100` and add it to `PAL_INITIALIZE_CORECLR` set of flags in the same file.

harisokanovic replied:
Pushed a new rev with your suggestion

harisokanovic committed:
pal init: InitializeFlushProcessWriteBuffers() before first thread to improve start time

InitializeFlushProcessWriteBuffers() initializes expedited membarrier()
syscall on Linux, which is much slower when called in a multi-thread
process. Move this init earlier to improve dotnet process start time.
A detailed explanation can be found in issue 106722.

Fixes https://github.com/dotnet/runtime/issues/106722

wash-amzn commented on src/coreclr/pal/inc/pal.h:
```diff
@@ -191,6 +191,7 @@ PAL_IsDebuggerPresent();
 #define PAL_INITIALIZE_ENSURE_STACK_SIZE            0x20
 #define PAL_INITIALIZE_REGISTER_SIGNALS             0x40
 #define PAL_INITIALIZE_REGISTER_ACTIVATION_SIGNAL   0x80
+#define PAL_INITIALIZE_FLUSH_PROCES_WRITE_BUFFERS   0x100
```
s/PROCES/PROCESS/

jkotas reviewed (approved):
Looks great! Thank you

Could you please accept the Contributor License Agreement?

harisokanovic commented:
@dotnet-policy-service agree [company="Amazon"]

harisokanovic commented:
@dotnet-policy-service agree company="Amazon"

jkotas commented:
Let me check why the bot does not like the command

harisokanovic commented:
> Let me check why the bot does not like the command

I think it worked - see `license/cla — All CLA requirements met` under approvals. I deleted my first reply which had a typo.

harisokanovic commented:
The test failures appear to be intermittent per [#100558](https://github.com/dotnet/runtime/issues/100558) and [#2496](https://github.com/dotnet/dnceng/issues/2496). I don't have permission to rerun those jobs.

harisokanovic commented:
@jkotas, is it possible to backport this change to dotnet8?
CC @Beau-Gosse-dev

jkotas commented:
/backport to release/9.0

jkotas commented:
> is it possible to backport this change to dotnet8?

I can backport it to .NET 9 RC for now.

This change does not meet our default servicing bar. It is a perf fix that is not fixing a regression, the problem existed for number of years since membarrier was introduced. Given that the fix is very simple and it provides measurable improvement for every app, I can ask for an exception for .NET 8 backport after the fix bakes for several months in main and .NET 9.
